File: tests/t_ethereum_evm_precompiles.nim
# Constantine
# Copyright (c) 2018-2019    Status Research & Development GmbH
# Copyright (c) 2020-Present Mamy André-Ratsimbazafy
# Licensed and distributed under either of
#   * MIT license (license terms in the root directory or at http://opensource.org/licenses/MIT).
#   * Apache v2 license (license terms in the root directory or at http://www.apache.org/licenses/LICENSE-2.0).
# at your option. This file may not be copied, modified, or distributed except according to those terms.

import
  # Standard library
  std/[os, strutils],
  # 3rd party
  pkg/jsony,
  # Internals
  constantine/serialization/codecs,
  constantine/ethereum_evm_precompiles

# Test vector source:
# - https://github.com/ethereum/go-ethereum/tree/release/1.14/core/vm/testdata/precompiles
# - https://github.com/ethereum/EIPs/tree/3b5fcad/assets/eip-2537

type
  HexString = string

  PrecompileTest = object
    Input: HexString
    Expected: HexString
    ExpectedError: string
    Name: string
    Gas: int
    NoBenchmark: bool

const
  TestVectorsDir* =
    currentSourcePath.rsplit(DirSep, 1)[0] / "protocol_ethereum_evm_precompiles"

proc loadVectors(TestType: typedesc, filename: string): TestType =
  let content = readFile(TestVectorsDir/filename)
  result = content.fromJson(TestType)

template runPrecompileTests(filename: string, funcname: untyped, outsize: int) =
  block:
    proc `PrecompileTestrunner _ funcname`() =
      let vec = seq[PrecompileTest].loadVectors(filename)
      echo "Running ", filename

      for test in vec:
        stdout.write "    Testing " & test.Name & " ... "

        # Length: 2 hex characters -> 1 byte
        var inputbytes = newSeq[byte](test.Input.len div 2)
        inputbytes.paddedFromHex(test.Input, bigEndian)

        var expected = newSeq[byte](test.Expected.len div 2)
        expected.paddedFromHex(test.Expected, bigEndian)

        ## TODO: change to use `modexp_result_size` API after rebase
        let outs = if outsize > 0: outsize else: test.Expected.len div 2
        var r = newSeq[byte](outs)

        let status = funcname(r, inputbytes)
        if status != cttEVM_Success:
          doAssert test.ExpectedError.len > 0, "[Test Failure]\n" &
            "  " & test.Name & "\n" &
            "  " & funcname.astToStr & "\n" &
            "  " & "Nim proc returned failure, but test expected to pass.\n" &
            "  " & "Expected result: " & $expected.toHex()
        else:
          doAssert r == expected, "[Test Failure]\n" &
            "  " & test.Name & "\n" &
            "  " & funcname.astToStr & " status: " & $status & "\n" &
            "  " & "result:   " & r.toHex() & "\n" &
            "  " & "expected: " & expected.toHex() & '\n'

        stdout.write "Success\n"

    `PrecompileTestrunner _ funcname`()

proc testSha256() =
  # https://github.com/ethereum/go-ethereum/blob/v1.14.0/core/vm/contracts_test.go#L206-L214
  let input = "38d18acb67d25c8bb9942764b62f18e17054f66a817bd4295423adf9ed98873e000000000000000000000000000000000000000000000000000000000000001b38d18acb67d25c8bb9942764b62f18e17054f66a817bd4295423adf9ed98873e789d1dd423d25f0772d2748d60f7e4b81bb14d086eba8e8e8efb6dcff8a4ae02"
  let expected = "811c7003375852fabd0d362e40e68607a12bdabae61a7d068fe5fdd1dbbf2a5d"

  echo "Running SHA256 tests"
  stdout.write "    Testing SHA256 ... "

  var inputbytes = newSeq[byte](input.len div 2)
  inputbytes.fromHex(input)

  var expectedbytes = newSeq[byte](expected.len div 2)
  expectedbytes.fromHex(expected)

  var r = newSeq[byte](expected.len div 2)

  let status = eth_evm_sha256(r, inputbytes)
  if status != cttEVM_Success:
    reset(r)

  doAssert r == expectedbytes, "[Test Failure]\n" &
    "  eth_evm_sha256 status: " & $status & "\n" &
    "  " & "result:   " & r.toHex() & "\n" &
    "  " & "expected: " & expectedbytes.toHex() & '\n'

  stdout.write "Success\n"

# ----------------------------------------------------------------------

testSha256()

runPrecompileTests("modexp.json", eth_evm_modexp, 0)
runPrecompileTests("modexp_eip2565.json", eth_evm_modexp, 0)

runPrecompileTests("bn256Add.json", eth_evm_bn254_g1add, 64)
runPrecompileTests("bn256ScalarMul.json", eth_evm_bn254_g1mul, 64)
runPrecompileTests("bn256Pairing.json", eth_evm_bn254_ecpairingcheck, 32)

runPrecompileTests("eip-2537/add_G1_bls.json", eth_evm_bls12381_g1add, 128)
runPrecompileTests("eip-2537/fail-add_G1_bls.json", eth_evm_bls12381_g1add, 128)
runPrecompileTests("eip-2537/add_G2_bls.json", eth_evm_bls12381_g2add, 256)
runPrecompileTests("eip-2537/fail-add_G2_bls.json", eth_evm_bls12381_g2add, 256)

runPrecompileTests("eip-2537/mul_G1_bls.json", eth_evm_bls12381_g1mul, 128)
runPrecompileTests("eip-2537/fail-mul_G1_bls.json", eth_evm_bls12381_g1mul, 128)
runPrecompileTests("eip-2537/mul_G2_bls.json", eth_evm_bls12381_g2mul, 256)
runPrecompileTests("eip-2537/fail-mul_G2_bls.json", eth_evm_bls12381_g2mul, 256)

runPrecompileTests("eip-2537/multiexp_G1_bls.json", eth_evm_bls12381_g1msm, 128)
runPrecompileTests("eip-2537/fail-multiexp_G1_bls.json", eth_evm_bls12381_g1msm, 128)
runPrecompileTests("eip-2537/multiexp_G2_bls.json", eth_evm_bls12381_g2msm, 256)
runPrecompileTests("eip-2537/fail-multiexp_G2_bls.json", eth_evm_bls12381_g2msm, 256)

runPrecompileTests("eip-2537/pairing_check_bls.json", eth_evm_bls12381_pairingcheck, 32)
runPrecompileTests("eip-2537/fail-pairing_check_bls.json", eth_evm_bls12381_pairingcheck, 32)

runPrecompileTests("eip-2537/map_fp_to_G1_bls.json", eth_evm_bls12381_map_fp_to_g1, 128)
runPrecompileTests("eip-2537/fail-map_fp_to_G1_bls.json", eth_evm_bls12381_map_fp_to_g1, 128)
runPrecompileTests("eip-2537/map_fp2_to_G2_bls.json", eth_evm_bls12381_map_fp2_to_g2, 256)
runPrecompileTests("eip-2537/fail-map_fp2_to_G2_bls.json", eth_evm_bls12381_map_fp2_to_g2, 256)

File: constantine-rust/constantine-ethereum-evm-precompiles/tests/t_ethereum_evm_precompiles.rs
//! Constantine
//! Copyright (c) 2018-2019    Status Research & Development GmbH
//! Copyright (c) 2020-Present Mamy André-Ratsimbazafy
//! Licensed and distributed under either of
//!   * MIT license (license terms in the root directory or at http://opensource.org/licenses/MIT).
//!   * Apache v2 license (license terms in the root directory or at http://www.apache.org/licenses/LICENSE-2.0).
//! at your option. This file may not be copied, modified, or distributed except according to those terms.

use constantine_ethereum_evm_precompiles::*;
use constantine_sys::ctt_evm_status;

use std::fs;

use hex;
use serde::Deserialize;
use serde_json;

// Rust does not support concatenating
// compile-time &str ¯\_(ツ)_/¯, so we need to use macros, C-style.

macro_rules! test_dir {
    () => {
        "../../tests/protocol_ethereum_evm_precompiles/"
    };
}

const MODEXP_TESTS: &str = concat!(test_dir!(), "modexp.json");
const MODEXP_EIP2565_TESTS: &str = concat!(test_dir!(), "modexp_eip2565.json");

const BN256ADD_TESTS: &str = concat!(test_dir!(), "bn256Add.json");
const BN256SCALARMUL_TESTS: &str = concat!(test_dir!(), "bn256ScalarMul.json");
const BN256PAIRING_TESTS: &str = concat!(test_dir!(), "bn256Pairing.json");

const ADD_G1_BLS_TESTS: &str = concat!(test_dir!(), "eip-2537/add_G1_bls.json");
const FAIL_ADD_G1_BLS_TESTS: &str = concat!(test_dir!(), "eip-2537/fail-add_G1_bls.json");
const ADD_G2_BLS_TESTS: &str = concat!(test_dir!(), "eip-2537/add_G2_bls.json");
const FAIL_ADD_G2_BLS_TESTS: &str = concat!(test_dir!(), "eip-2537/fail-add_G2_bls.json");

const MUL_G1_BLS_TESTS: &str = concat!(test_dir!(), "eip-2537/mul_G1_bls.json");
const FAIL_MUL_G1_BLS_TESTS: &str = concat!(test_dir!(), "eip-2537/fail-mul_G1_bls.json");
const MUL_G2_BLS_TESTS: &str = concat!(test_dir!(), "eip-2537/mul_G2_bls.json");
const FAIL_MUL_G2_BLS_TESTS: &str = concat!(test_dir!(), "eip-2537/fail-mul_G2_bls.json");

const MULTIEXP_G1_BLS_TESTS: &str = concat!(test_dir!(), "eip-2537/multiexp_G1_bls.json");
const FAIL_MULTIEXP_G1_BLS_TESTS: &str = concat!(test_dir!(), "eip-2537/fail-multiexp_G1_bls.json");
const MULTIEXP_G2_BLS_TESTS: &str = concat!(test_dir!(), "eip-2537/multiexp_G2_bls.json");
const FAIL_MULTIEXP_G2_BLS_TESTS: &str = concat!(test_dir!(), "eip-2537/fail-multiexp_G2_bls.json");

const PAIRING_CHECK_BLS_TESTS: &str = concat!(test_dir!(), "eip-2537/pairing_check_bls.json");
const FAIL_PAIRING_CHECK_BLS_TESTS: &str =
    concat!(test_dir!(), "eip-2537/fail-pairing_check_bls.json");

const MAP_FP_TO_G1_BLS_TESTS: &str = concat!(test_dir!(), "eip-2537/map_fp_to_G1_bls.json");
const FAIL_MAP_FP_TO_G1_BLS_TESTS: &str =
    concat!(test_dir!(), "eip-2537/fail-map_fp_to_G1_bls.json");
const MAP_FP2_TO_G2_BLS_TESTS: &str = concat!(test_dir!(), "eip-2537/map_fp2_to_G2_bls.json");
const FAIL_MAP_FP2_TO_G2_BLS_TESTS: &str =
    concat!(test_dir!(), "eip-2537/fail-map_fp2_to_G2_bls.json");

type HexString = String;

#[allow(dead_code, non_snake_case)]
#[derive(Deserialize, Debug)]
struct PrecompileTest {
    Input: HexString,
    #[serde(default)]
    Expected: Option<HexString>,
    #[serde(default)]
    ExpectedError: Option<String>,
    Name: String,
    #[serde(default)]
    Gas: i64,
    #[serde(default)]
    NoBenchmark: bool,
}

fn from_hex(hex: HexString) -> Option<Vec<u8>> {
    // data does not always have `0x` prefix in JSON files!
    // Check for the '0x' prefix
    let href: &[u8] = hex.as_ref();
    let href = if href.starts_with(b"0x") {
        &href[2..]
    } else {
        href
    };
    // TODO: why cannot use `with_capacity`?
    let mut result = vec![0u8; href.len() / 2];
    match hex::decode_to_slice(href, result.as_mut_slice() as &mut [u8]) {
        Ok(_) => Some(result),
        Err(_) => None,
    }
}

fn t_generate<T, F>(test_name: String, func: F)
where
    T: IntoIterator<Item = u8>,
    F: Fn(&[u8]) -> Result<T, ctt_evm_status>,
{
    type TestVectors = Vec<PrecompileTest>;

    let unparsed = fs::read_to_string(&test_name).unwrap();
    let vectors: TestVectors = serde_json::from_str(&unparsed).expect(&format!(
        "Formatting should be consistent for file \"{}\"",
        &test_name
    ));

    for vector in vectors {
        println!("Running test case: {}", vector.Name);

        let input = vector.Input;
        let expected = vector.Expected;

        let input_bytes =
            from_hex(input).expect("Test failed; input bytes could not be unmarshaled.");
        // Call the test function
        let result = func(&input_bytes);
        match result {
            Ok(r) => {
                assert!(expected.is_some());
                let expected_bytes = from_hex(expected.unwrap());
                assert!(r.into_iter().collect::<Vec<u8>>() == expected_bytes.unwrap());
            }
            Err(_) => {
                // in case of error there must not be an `expected` field
                assert!(expected.is_none());
                let expected_error = vector.ExpectedError;
                assert!(expected_error.is_some());
            }
        };
    }
}

#[test]
fn t_modexp() {
    let test_name = MODEXP_TESTS.to_string();
    t_generate(test_name, evm_modexp);
}

#[test]
fn t_modexp_eip2565_tests() {
    let test_name = MODEXP_EIP2565_TESTS.to_string();
    t_generate(test_name, evm_modexp);
}

#[test]
fn t_bn256add_tests() {
    let test_name = BN256ADD_TESTS.to_string();
    t_generate(test_name, evm_bn254_g1add);
}
#[test]
fn t_bn256scalarmul_tests() {
    let test_name = BN256SCALARMUL_TESTS.to_string();
    t_generate(test_name, evm_bn254_g1mul);
}

#[test]
fn t_bn256pairing_tests() {
    let test_name = BN256PAIRING_TESTS.to_string();
    t_generate(test_name, evm_bn254_ec_pairing_check);
}

#[test]
fn t_add_g1_bls_tests() {
    let test_name = ADD_G1_BLS_TESTS.to_string();
    t_generate(test_name, evm_bls12381_g1add);
}
#[test]
fn t_fail_add_g1_bls_tests() {
    let test_name = FAIL_ADD_G1_BLS_TESTS.to_string();
    t_generate(test_name, evm_bls12381_g1add);
}
#[test]
fn t_add_g2_bls_tests() {
    let test_name = ADD_G2_BLS_TESTS.to_string();
    t_generate(test_name, evm_bls12381_g2add);
}
#[test]
fn t_fail_add_g2_bls_tests() {
    let test_name = FAIL_ADD_G2_BLS_TESTS.to_string();
    t_generate(test_name, evm_bls12381_g2add);
}

#[test]
fn t_mul_g1_bls_tests() {
    let test_name = MUL_G1_BLS_TESTS.to_string();
    t_generate(test_name, evm_bls12381_g1mul);
}
#[test]
fn t_fail_mul_g1_bls_tests() {
    let test_name = FAIL_MUL_G1_BLS_TESTS.to_string();
    t_generate(test_name, evm_bls12381_g1mul);
}
#[test]
fn t_mul_g2_bls_tests() {
    let test_name = MUL_G2_BLS_TESTS.to_string();
    t_generate(test_name, evm_bls12381_g2mul);
}
#[test]
fn t_fail_mul_g2_bls_tests() {
    let test_name = FAIL_MUL_G2_BLS_TESTS.to_string();
    t_generate(test_name, evm_bls12381_g2mul);
}

#[test]
fn t_multiexp_g1_bls_tests() {
    let test_name = MULTIEXP_G1_BLS_TESTS.to_string();
    t_generate(test_name, evm_bls12381_g1msm);
}
#[test]
fn t_fail_multiexp_g1_bls_tests() {
    let test_name = FAIL_MULTIEXP_G1_BLS_TESTS.to_string();
    t_generate(test_name, evm_bls12381_g1msm);
}
#[test]
fn t_multiexp_g2_bls_tests() {
    let test_name = MULTIEXP_G2_BLS_TESTS.to_string();
    t_generate(test_name, evm_bls12381_g2msm);
}
#[test]
fn t_fail_multiexp_g2_bls_tests() {
    let test_name = FAIL_MULTIEXP_G2_BLS_TESTS.to_string();
    t_generate(test_name, evm_bls12381_g2msm);
}

#[test]
fn t_pairing_check_bls_tests() {
    let test_name = PAIRING_CHECK_BLS_TESTS.to_string();
    t_generate(test_name, evm_bls12381_pairing_check);
}
#[test]
fn t_fail_pairing_check_bls_tests() {
    let test_name = FAIL_PAIRING_CHECK_BLS_TESTS.to_string();
    t_generate(test_name, evm_bls12381_pairing_check);
}

#[test]
fn t_map_fp_to_g1_bls_tests() {
    let test_name = MAP_FP_TO_G1_BLS_TESTS.to_string();
    t_generate(test_name, evm_bls12381_map_fp_to_g1);
}
#[test]
fn t_fail_map_fp_to_g1_bls_tests() {
    let test_name = FAIL_MAP_FP_TO_G1_BLS_TESTS.to_string();
    t_generate(test_name, evm_bls12381_map_fp_to_g1);
}
#[test]
fn t_map_fp2_to_g2_bls_tests() {
    let test_name = MAP_FP2_TO_G2_BLS_TESTS.to_string();
    t_generate(test_name, evm_bls12381_map_fp2_to_g2);
}
#[test]
fn t_fail_map_fp2_to_g2_bls_tests() {
    let test_name = FAIL_MAP_FP2_TO_G2_BLS_TESTS.to_string();
    t_generate(test_name, evm_bls12381_map_fp2_to_g2);
}
